lib/wallet/http.js

```diff
@@ -1427,6 +1427,99 @@ class HTTP extends Server {
+
+      assert(names && names.length > 0, 'Names are required.');
+
+      names.forEach((element) => {
+        const validator = new Validator(element);
+        if (!validator.has('name') || !validator.has('data'))
+           throw new Error('name and data must be present in every element.');
+
+        validator.str('name');
+        validator.obj('data');
+
+        const dataValidator = new Validator(element.data);
+        if (!dataValidator.has('records'))
+           throw new Error('data must be in form: {records: []}');
+
+        validator.array(element.data.records);
+      });
+
+      names = names.map(({name, data}) => {
+        return {name, data: Resource.fromJSON(data)};
+      });
+
+      const options = TransactionOptions.fromValidator(valid);
+
+      const finishCache = req.wallet.sendFinishResults;
+
+      const processedFinishes = [];
+      let errorMessages = [];
+      const uniqueNames = [];
+
+      names.forEach(({name, data}) => {
+        const cachedNames = finishCache.get(name);
+        if (!cachedNames) {
+          uniqueNames.push({name, data});
+        } else {
+          cachedNames.forEach(cachedName => processedFinishes.push(cachedName));
+        }
+      });
+
+      if (uniqueNames.length > 0) {
+        const nameHashMap = new Map();
+        uniqueNames.forEach(({name, data}) => {
+          const nameHash = rules.hashName(name).toString('hex');
+          nameHashMap.set(nameHash, name);
+        });
+
+        // always broadcast
+        const {mtx, errorMessages: errors} = await req.wallet
+           .createBatchFinish(uniqueNames, options);
+
+        errorMessages = errors;
+        const tx = await req.wallet.sendMTX(mtx, passphrase);
+        const txJSON = tx.toJSON(this.network);
+
+        const outputsLength = txJSON.outputs.length;
+        for (let i=0; i<outputsLength; i++) {
+          const output = txJSON.outputs[i];
+          const covenant = output.covenant;
+          const covenantType = covenant.type;
+
+          if (covenantType === rules.types.REDEEM ||
+              covenantType === rules.types.REGISTER) {
+                const nameHash = covenant.items[0];
+                const name = nameHashMap.get(nameHash);
+                const item = {
+                  idempotency_key: name,
+                  tx_hash: txJSON.hash,
+                  output_index: i,
+                  output: output,
+                  from_cache: false
+                };
+                processedFinishes.push(item);
+                // populate cache
+                const cachedItem = Object.assign({}, item);
+                cachedItem.from_cache = true;
+                // Handle Cache Insert
+                if (!finishCache.get(name)) {
+                  finishCache.set(name, []);
+                }
+
+                finishCache.get(name).push(cachedItem);
+              }
+          }
+      }
+
+      return res.json(200, {processedFinishes,errorMessages});
+    });
```

Comment:
meta code quality comment: this function got quite large, is there already a standard place for helper functions for this wallet/http code? Would be a good idea to break this up into more pieces